File: solutions/clojure/kindergarten-garden/1/src/kindergarten_garden.clj
(ns kindergarten-garden
  (:require [clojure.string :as st]))

(def ->plants
  {\R :radishes
   \C :clover
   \G :grass
   \V :violets})

(def children
  ["alice" "bob" "charlie" "david" "eve"
   "fred" "ginny" "harriet" "ileana"
   "joseph" "kincaid" "larry"])

(def map-plants (partial map ->plants))

(def ->name (comp keyword st/lower-case))

(defn garden
  ([plants] (garden plants children))
  ([plants students]
   (let [children (sort (map ->name students))]
     (->> plants
          st/split-lines
          (map (partial partition 2))
          (apply map (comp map-plants flatten vector))
          (zipmap children)))))

File: solutions/csharp/kindergarten-garden/1/KindergartenGarden.cs
public enum Plant
{
    Violets,
    Radishes,
    Clover,
    Grass
}

public class KindergartenGarden(string diagram)
{
    private readonly string[] rows = diagram.Split("\n");

    private enum Student
    {
        Alice, Bob, Charlie, David, Eve, Fred,
        Ginny, Harriet, Ileana, Joseph, Kincaid, Larry
    }

    public IEnumerable<Plant> Plants(string student)
    {
        var offset = 2 * (int)Enum.Parse<Student>(student);
        return rows
            .SelectMany(row => row.Skip(offset).Take(2))
            .Select(ToPlant);
    }

    private static Plant ToPlant(char code) => code switch
    {
        'V' => Plant.Violets,
        'R' => Plant.Radishes,
        'C' => Plant.Clover,
        'G' => Plant.Grass,
        _ => throw new ArgumentException("Unknown plant code.", nameof(code))
    };
}
